src/kaboom.ts

```diff
@@ -5467,11 +5477,335 @@ export default (gopt: KaboomOpt = {}): KaboomCtx => {
+				const costSoFar = new Map<number, number>()
+				costSoFar.set(start, 0)
+	
+				while (frontier.length !== 0) {
+					const current = frontier.remove()?.node
+	
+					if (current === goal)
+						break
+	
+					const neighbours = this._getNeighbours(current, diagonals)
+					for (const next of neighbours) {
+						const newCost = (costSoFar.get(current) || 0) +
+							this._getCost(current, next) +
+							this._getHeuristic(next, goal)
+						if (!costSoFar.has(next) || newCost < costSoFar.get(next)) {
+							costSoFar.set(next, newCost)
+							frontier.insert({ cost: newCost, node: next })
+							cameFrom.set(next, current)
+						}
+					}
+				}
+	
+				const path = []
+				let node = goal
+				const p = this._hash2Tile(node)
+				path.push(p)
+				while (node !== start) {
+					node = cameFrom.get(node)
+					const p = this._hash2Tile(node)
+					path.push(p)
+				}
+				return path.reverse()
+			},
+	
+			getPath(this: GameObj<LevelComp | PathfindingComp>, from: Vec2, to: Vec2, diagonals: boolean) {
+				const tw = this.tileWidth()
+				const th = this.tileHeight()
+				const path = this.getTilePath(this.pos2Tile(from), this.pos2Tile(to), diagonals)
+				if (path) {
+					return [from, ...path.slice(1, -1).map(tilePos => tilePos.scale(tw, th).add(tw / 2, th / 2)), to]
+				}
+				else {
+					return null
+				}
+			},
+	
+			/**
+			 * Private methods start here
+			 */
+	
+			_getNeighbours(this: GameObj<LevelComp | PathfindingComp>, node: number, diagonals?: boolean) {
+				const numColumns = this.numColumns()
+				const numRows = this.numRows()
+				const n = []
+				const x = Math.floor(node % numColumns)
+				const left = x > 0 &&
+					(_edgeMap[node] & EdgeMask.Left) &&
+					_costMap[node - 1] !== Infinity
+				const top = node >= numColumns &&
+					(_edgeMap[node] & EdgeMask.Top) &&
+					_costMap[node - numColumns] !== Infinity
+				const right = x < numColumns - 1 &&
+					(_edgeMap[node] & EdgeMask.Right) &&
+					_costMap[node + 1] !== Infinity
+				const bottom = node < numColumns * numRows - numColumns - 1 &&
+					(_edgeMap[node] & EdgeMask.Bottom) &&
+					_costMap[node + numColumns] !== Infinity
+				if (diagonals) {
+					if (left) {
+						if (top) { n.push(node - numColumns - 1) }
+						n.push(node - 1)
+						if (bottom) { n.push(node + numColumns - 1) }
+					}
+					if (top) {
+						n.push(node - numColumns)
+					}
+					if (right) {
+						if (top) { n.push(node - numColumns + 1) }
+						n.push(node + 1)
+						if (bottom) { n.push(node + numColumns + 1) }
+					}
+					if (bottom) {
+						n.push(node + numColumns)
+					}
+				}
+				else {
+					if (left) {
+						n.push(node - 1)
+					}
+					if (top) {
+						n.push(node - numColumns)
+					}
+					if (right) {
+						n.push(node + 1)
+					}
+					if (bottom) {
+						n.push(node + numColumns)
+					}
+				}
+				return n
+			},
+	
+			_getCost(this: GameObj<LevelComp | PathfindingComp>, node: number, neighbour: number) {
```

Comment:
Is `node` parameter unused here?

Reply:
Currently yes, because we only take the destination cost into account. In some graphs you may take both source and destination into account. For example if you have a height map, 1 to 6 means a slope of 5 upwards. In other cases the cost may be stored in edges of the graph rather than the nodes, so you might need both nodes to index the edge.